sc-memory/sc-memory/sc_agent_context.hpp

```diff
@@ -0,0 +1,336 @@
+   * @warning Specified agent class must be derived from class `ScAgent`.
+   *  @throws utils::ExceptionInvalidParams if any of the subscription addresses are invalid.
+   * @throws utils::ExceptionInvalidState if the agent is not subscribed to the event.
+   */
+  template <class TScAgent, class... TScAddr>
+  _SC_EXTERN typename std::enable_if<!std::is_base_of<ScActionAgent, TScAgent>::value>::type UnsubscribeAgent(
+      TScAddr const &... subscriptionAddrs);
+
+  /*!
+   * @brief Unsubscribes agent class from sc-event of adding outgoing sc-arc from `action_initiated` to some formed
+   * sc-action.
+   * @tparam TScAgent An agent class to be subscribed to the event.
+   * @warning Specified agent class must be derived from class `ScAgentAction`.
+   * @throws utils::ExceptionInvalidParams if any of the subscription addresses are invalid.
+   * @throws utils::ExceptionInvalidState if the agent is not subscribed to the event.
+   */
+  template <class TScAgent>
+  _SC_EXTERN typename std::enable_if<std::is_base_of<ScActionAgent, TScAgent>::value>::type UnsubscribeAgent();
+
+  /*!
+   * @brief Loads specification of agent implementation and subscribes agent class.
+   * @tparam TScAgent An agent class to be subscribed to the event.
+   * @param agentImplementationAddr A sc-address of agent implementation specified in knowledge base for this agent
+   * class.
+   * @throws utils::ExceptionInvalidParams if any of the subscription addresses are invalid.
+   * @throws utils::ExceptionInvalidState if the agent is already subscribed to the event.
+   * @throws utils::ExceptionInvalidState if the agent implementation for this agent class is valid and is not included
+   * in any abstract sc-agent.
+   * @throws utils::ExceptionInvalidState if the agent implementation for this agent class is valid and the abstract
+   * sc-agent for this agent class does not have a primary initiation condition.
+   */
+  template <class TScAgent>
+  _SC_EXTERN void LoadAndSubscribeAgent(ScAddr const & agentImplementationAddr);
+
+  /*!
+   * @brief Remove from program specification of agent and unsubscribes agent class.
+   * @tparam TScAgent An agent class to be unsubscribed from the event.
+   * @param agentImplementationAddr A sc-address of agent implementation specified in knowledge base for this agent
+   * class.
+   * @throws utils::ExceptionInvalidParams if any of the subscription addresses are invalid.
+   * @throws utils::ExceptionInvalidState if the agent is not subscribed to the event.
+   * @throws utils::ExceptionInvalidState if the agent implementation for this agent class is valid and is not included
+   * in any abstract sc-agent.
+   * @throws utils::ExceptionInvalidState if the agent implementation for this agent class is valid and the abstract
+   * sc-agent for this agent class does not have a primary initiation condition.
+   */
+  template <class TScAgent>
+  _SC_EXTERN void DestroyAndUnsubscribeAgent(ScAddr const & agentImplementationAddr);
```

Comment:
I couldn't find any destroys in this method, only unsubscribe

Reply:
I wanted that they will be named `SubscribeAgent` and `UnsubscribeAgent`. But there is a problem. The existing `SubscribeAgent` takes `ScAddr`, and `LoadAndSubscribeAgent` takes also `ScAddr`. I can't name them any other way

Reply:
The same situation for `UnsubscribeAgent` and `DestroyAndUnsubscribeAgent`

Reply:
what about SubscribeAgentImplementation and SubscribeAgent?

Reply:
Let's discuss it